### ISSUE TITLE
Code snippet in 8.2.6. Error in CPL_transform(x, crs, aoi, pipeline, reverse) :    OGRCreateCoordinateTransformation() returned NULL: PROJ available?

### ISSUE THREAD

**Issue by Linlin-Sun** [Robinlovelace/geocompr#497]

**Comment by Linlin-Sun** [comment]
I got "Error in data.matrix(data) : 
  'list' object cannot be coerced to type 'double'" running below line but was able to get round it. 

urb_1970_2030 = urban_agglomerations %>% 
    filter(year %in% c(1970, 1990, 2010, 2030))

I got "Error in CPL_transform(x, crs, aoi, pipeline, reverse) : 
  OGRCreateCoordinateTransformation() returned NULL: PROJ available?" and not sure how to make it work. 
Running two tm_shape() separately works for both but got the error when combining them together. 
tm_shape(world) +
    tm_polygons() +
    tm_shape(urb_1970_2030) +
    tm_symbols(col = "black", border.col = "white", size = "population_millions") +
    tm_facets(by = "year", nrow = 2, free.coords = FALSE)

**Comment by Robinlovelace** [comment]
Thanks for reporting. What operating system are you on and what versions of GDAL, GEOS and PROJ are you running? Those details may help identify the problem. If you can provide a reproducible example that will also help.

You can find out what versions of those libraries you have with the following command:

``` r
library(sf)
#> Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 7.0.0
```

<sup>Created on 2020-04-13 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by rovisope** [comment]
I have found the same problem, I am working with MacOs Catalina 10.15.4. 

When I wrote "library(sf)" I can't see any message below. But running "sf_extSoftVersion()" I can see:
          GEOS           GDAL         proj.4 GDAL_with_GEOS     USE_PROJ_H 
       "3.7.2"        "2.4.2"        "5.2.0"        "false"        "false" 

I was trying to run the code for the "FIGURE 8.13: Faceted map showing the top 30 largest urban agglomerations from 1970 to 2030 based on population projections by the United Nations." so: 

urb_1970_2030 = urban_agglomerations %>% 
  filter(year %in% c(1970, 1990, 2010, 2030))

tm_shape(world) +
  tm_polygons() +
  tm_shape(urb_1970_2030) +
  tm_symbols(col = "black", border.col = "white", size = "population_millions") +
  tm_facets(by = "year", nrow = 2, free.coords = FALSE)

Thank you so much for everything. Hopefully this information helps!

**Comment by rovisope** [comment]
BTW, the following similar code works in my computer:

data(World, metro)

m2 <- tm_shape(World, simplify = 0.5) +
  tm_fill() +
  tm_shape(metro) + 
  tm_bubbles(size = paste0("pop", seq(1970, 2030, by=10)),
             col = "purple",
             border.col = "black", border.alpha = .5,
             scale = 2) +
  tm_facets(free.scales.symbol.size = FALSE, nrow=1,ncol=1) + 
  tm_format("World", scale=.5)

Thank you again. Hope it helps!

**Comment by Robinlovelace** [comment]
Hi @rovisope, thanks for the clarification. I have encountered similar error messages, including on our automated test build system: https://travis-ci.org/github/Robinlovelace/geocompr/jobs/673715672#L6119

Possible solution: try re-installing the `sf` package as follows:

```r
install.packages("sf")
```

and if that fails try the development version:

```r
remotes::install_github("r-spatial/sf")
```

Let me know how you get on. If it still fails it may be due to incompatibility between your version of PROJ and your version of `sf`.

**Comment by Robinlovelace** [comment]
Also try installing the latest version of `tmap`, that could help...

**Comment by rovisope** [comment]
Thank you so much @Robinlovelace!

The problem persists after reinstalling sf and tmaps... 

I could not install the development version of sf (It says "configure: error: gdal-config not found or not executable.".)

I am new on R and geocomputation... I really do not know how to make compatible PROJ and sf...

**Comment by Robinlovelace** [comment]
Can you try installing the binary version? I am not sure. I suggest taking a look for similar issues here: https://github.com/r-spatial/sf/search?q=mac+install&type=Issues

Please do not open an issue on the `sf` issue tracker unless you're sure you have found a new issue, you cannot solve the problem using information from previous issues, and you have a very clear reproducible example. @edzer has been dealing with lots of tricky install problems so no need to add to open another issue I suspect.

Hope this guidance is useful, let us know how you get on.

**Comment by Linlin-Sun** [comment]
> Thanks for reporting. What operating system are you on and what versions of GDAL, GEOS and PROJ are you running? Those details may help identify the problem. If you can provide a reproducible example that will also help.
> 
> You can find out what versions of those libraries you have with the following command:
> 
> ```r
> library(sf)
> #> Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 7.0.0
> ```
> 
> Created on 2020-04-13 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)

Hi, Robin, Thanks for your response. 
I got below after following your instruction. 
> library(sf)
Linking to GEOS 3.6.1, GDAL 2.2.3, PROJ 4.9.3

**Comment by Robinlovelace** [comment]
Thanks @Linlin-Sun, but what OS are you running?

**Comment by Linlin-Sun** [comment]
> Thanks @Linlin-Sun, but what OS are you running?

I am on windows 10.

**Comment by Robinlovelace** [comment]
> I am on windows 10.

Have you tried re-installing the `sf` package?

**Comment by rovisope** [comment]
Hi @Robinlovelace. Thanks for your support.

I have tried different solutions but I am still stuck. I could install the development session but now I found the following results:

> library(sf)
Linking to GEOS 3.7.2, GDAL 2.4.2, PROJ 5.2.0
> library(dplyr)

Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union

> library(spData)
> library(tmap)    # for static and interactive maps
> 
> 
> urb_1970_2030 = urban_agglomerations %>% 
+   filter(year %in% c(1970, 1990, 2010, 2030))
> 
> tm_shape(world) +
+   tm_polygons() +
+   tm_shape(urb_1970_2030) +
+   tm_symbols(col = "black", border.col = "white", size = "population_millions") +
+   tm_facets(by = "year", nrow = 2, free.coords = FALSE)
Error in CPL_transform(x, crs, aoi, pipeline, reverse) : 
  OGRCreateCoordinateTransformation() returned NULL: PROJ available?
In addition: Warning message:
In CPL_transform(x, crs, aoi, pipeline, reverse) :
  GDAL Error 1: No PROJ.4 translation for source SRS, coordinate transformation initialization has failed.

I have tried to install version 0.8-1 but I was not succesful. I will keep you posted if I found a solution in the future.

Best, and thanks again!

Marti

**Comment by edzer** [comment]
Lovely. Looks like PROJ 5.x doesn't accept the wkt produces by PROJ >= 6.x. After resetting the crs with
```r
st_crs(urb_1970_2030) = st_crs(4326)
```
your `tmap` stuff works. That, or upgrading your system to PROJ >= 6 and reinstalling sf.

**Comment by rovisope** [comment]
Great! The code works! Thank you so much! 

(BTW, in my terminal appears that I have installed PROJ 7.0.0. but in R is stuck to PROJ 5.2.0. - Just in case...)

**Comment by Robinlovelace** [comment]
Great to hear it works @rovisope, thanks for this piece of good news (we need it hear in the UK at the moment ; )

I *think* the underlying causes of the issue have been identified. It's clear that the work-around mentioned by Edzer can fix it. One question is whether we should include that fix in the book? I guess not but will leave this issue open for now in case there are others who stumble across similar issues.

**Comment by rovisope** [comment]
Thanks to all of you for helping! Lots of energy in these difficult times!

**Comment by Robinlovelace** [comment]
Have you tried the suggestions above @Linlin-Sun ? I think the issue is resolved. Another option is to use Docker to get recent versions of the software.

**Comment by phungdo** [comment]
> Lovely. Looks like PROJ 5.x doesn't accept the wkt produces by PROJ >= 6.x. After resetting the crs with
> 
> ```r
> st_crs(urb_1970_2030) = st_crs(4326)
> ```
> 
> your `tmap` stuff works. That, or upgrading your system to PROJ >= 6 and reinstalling sf.

Hello Edzer, where should I put those code in?
Sorry for asking such question but I am new to R. Thank you. I can run library(sf) and the command returns  

> Linking to GEOS 3.6.1, GDAL 2.2.3, PROJ 4.9.3

**Comment by Robinlovelace** [comment]
Thanks for the question @phungdo but you need to provide a reproducible example for us to be able to help.

**Comment by Nowosad** [comment]
I degraded CRS for the `urban_agglomerations` from PROJ7 to PROJ5. 
@phungdo try `remotes::install_github("nowosad/spData")` first (before the rest of the code).

**Comment by Nowosad** [comment]
Closing for now. Free to let us now if the issue is still there.
